provider/aws: implement CloudFront Lambda Function Associations (supersedes #10985)

catsby opened the pull request:
Continue the work done in #10985 with just a few tweaks to restore the `TypeSet` in `lambda_function_association`.

Should also fix #11295

stack72 commented:
Solid work - this LGTM! Nice to see this coming in - it should unblock some people :)

MrGossett commented:
Thanks!

MrGossett commented on builtin/providers/aws/cloudfront_distribution_configuration_structure_test.go:
```diff
@@ -44,6 +49,35 @@ func trustedSignersConf() []interface{} {
 	return []interface{}{"1234567890EX", "1234567891EX"}
 }
 
+func lambdaSetHash(v interface{}) int {
```
This is essentially where my PR was at [9c5a25d](https://github.com/hashicorp/terraform/pull/11291/files/9c5a25dd480e71a3b0838c4c2f3f1cae47f41757) on Jan 5. I was derailed by the [comment from @apparentlymart](https://github.com/hashicorp/terraform/pull/10985#discussion_r94830032) asking to remove the hash function.

Sorry this got delayed 2 weeks as a result 🤓

catsby replied:
Nah it's not your fault. 

@apparentlymart was correct in that the manually added `Set` function should be omitted, unless there's a really good reason. 

The problem here was reconstructing "sets" in _code_ in the test instead of configuration. All the previous examples I found in the codebase were simple set's of `strings` which is trivial to reconstruct, but `NewSet` is more difficult with sets that have a more complicated schema. 

Like I said you did 99% of the work here I just needed to do some battlefield surgery 😄

MrGossett replied:
Roger. Thanks again 😄

catsby replied:
@MrGossett So.... funny enough, I had to reinstate that lambda set hash you were asked to remove 😆  

From Martin's comment:

> We only really continue to support explicit Set for compatibility with schemas that predate the automatic hashing.

This is _true_, but unfortunately `lambda_function_association` is being added as a sub param to two params that do themselves have custom hash function for the IDs. 

So, I had to add `lambda_function_association` to the two `CacheBehaviorHash` (?) method(s), but to do so, I'd need to hash them... so I needed that method 😐 

So anyway it's re-added

MrGossett replied:
:+1:

catsby commented:
Ran some test:

```
=== RUN   TestAccAWSCloudFrontDistribution_importBasic
--- PASS: TestAccAWSCloudFrontDistribution_importBasic (1587.37s)
=== RUN   TestAccAWSCloudFrontDistribution_S3Origin
--- PASS: TestAccAWSCloudFrontDistribution_S3Origin (1626.38s)
=== RUN   TestAccAWSCloudFrontDistribution_S3OriginWithTags
--- PASS: TestAccAWSCloudFrontDistribution_S3OriginWithTags (1734.81s)
```

catsby commented:
Hold off on this - need to investigate something

catsby commented:
I added documentation and the `lambda_function_association` objects to the `cacheBehaviorHash` methods to track them, and added docs in https://github.com/hashicorp/terraform/pull/11291/commits/6d4f4007feb1d177f48d5483da4eba537ec44056

catsby commented:
Hey @stack72 when you get a chance can you look this over? Not sure about the TravisCI error 🤔

kalbasit commented:
```

tomelliff commented on website/source/docs/providers/aws/r/cloudfront_distribution.html.markdown:
```diff
@@ -173,10 +171,13 @@ of several sub-resources - these resources are laid out below.
+  * `forwarded_values` (Required) - The [forwarded values configuration](#forwarded-values-arguments) that specifies how CloudFront
     handles query strings, cookies and headers (maximum one).
 
+  * `lambda_function_association` (Optional) - A complext config that triggers a lambda function with
```
Typo on `complext`

tomelliff commented on website/source/docs/providers/aws/r/cloudfront_distribution.html.markdown:
```diff
@@ -222,6 +223,19 @@ of several sub-resources - these resources are laid out below.
+Lambda@Edge](http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/what-is-lambda-at-edge.html) 
+for more information
+
+  * `event_type` (Required) - The specific event to trigger this functionon.
```
Typo on `functionon`

stack72 commented:
@catsby the travis fail is because of this

```
--- FAIL: TestCloudFrontStructure_flattenDefaultCacheBehavior (0.00s)
panic: interface conversion: interface {} is []interface {}, not *schema.Set [recovered]
	panic: interface conversion: interface {} is []interface {}, not *schema.Set
```

stack72 commented:
LGTM! 
![image](https://cloud.githubusercontent.com/assets/227823/22159363/612fec1e-df38-11e6-8178-3a2dce20cca5.png)

MrGossett commented on website/source/docs/providers/aws/r/cloudfront_distribution.html.markdown:
```diff
@@ -222,6 +223,19 @@ of several sub-resources - these resources are laid out below.
+for more information
+
+  * `event_type` (Required) - The specific event to trigger this function.
+  Valid values: `viewwer-request`, `origin-request`, `viewer-response`,
```
typo: `viewwer-request`

catsby replied:
fixed, thanks!

MrGossett commented on builtin/providers/aws/cloudfront_distribution_configuration_structure.go:
```diff
@@ -294,6 +311,9 @@ func flattenCacheBehavior(cb *cloudfront.CacheBehavior) map[string]interface{} {
 		m["trusted_signers"] = flattenTrustedSigners(cb.TrustedSigners)
 	}
+	if len(cb.LambdaFunctionAssociations.Items) > 0 {
+		m["lambda_function_association"] = flattenLambdaFunctionAssociations(cb.LambdaFunctionAssociations)
```
is this supposed to be plural? `lambda_function_associations`

catsby replied:
no, it's a key in the schema. If it were plural then the config to use it would be 

```
lambda_funcation_associations {
  event_type = "blah"
  arn = "blah"
}
lambda_funcation_associations {
  event_type = "blah"
  arn = "blah"
}
...
```

MrGossett replied:
Makes sense.

MrGossett commented on builtin/providers/aws/cloudfront_distribution_configuration_structure_test.go:
```diff
@@ -337,6 +363,9 @@ func TestCloudFrontStructure_flattenCacheBehavior(t *testing.T) {
 	if out["target_origin_id"] != "myS3Origin" {
 		t.Fatalf("Expected out[target_origin_id] to be myS3Origin, got %v", out["target_origin_id"])
 	}
+	if reflect.DeepEqual(out["lambda_function_associations"], in["lambda_function_associations"]) != true {
```
I have a feeling this is passing because `reflect.DeepEqual` is comparing `nil` to `nil`. The singular version is used in [defaultCacheBehaviorConf](https://github.com/hashicorp/terraform/pull/11291/files#diff-58242079a8a0215c803ef662dacf67e4R19), so the plural `lambda_function_associations` doesn't exist.

catsby replied:
makes sense, I'll investigate

catsby replied:
fixed in fd3f4bbf3de1398cd6f834166053c2676baf5e59
